DatadogLogs/Sources/LoggerProtocol.swift

```diff
@@ -0,0 +1,253 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-Present Datadog, Inc.
+ */
+
+import Foundation
+import DatadogInternal
+
+/// Log levels ordered by their severity, with `.debug` being the least severe and
+/// `.critical` being the most severe.
+public enum LogLevel: Int, Codable {
+    case debug
+    case info
+    case notice
+    case warn
+    case error
+    case critical
+}
+
+/// Datadog Logger.
+///
+/// Usage:
+///
+///     import DatadogLogs
+///
+///     // Initialise the Logs module
+///
+///     // logger reference
+///     var logger = Logger.create()
+public protocol LoggerProtocol {
+    /// General purpose logging method.
+    /// Sends a log with certain `level`, `message`, `error` and `attributes`.
+    ///
+    /// Although it can be used directly, it is more convenient and recommended to use specific methods declared on `Logger`:
+    /// * `debug(_:error:attributes:)`
+    /// * `info(_:error:attributes:)`
+    /// * ...
+    ///
+    /// - Parameters:
+    ///   - level: the log level
+    ///   - message: the message to be logged
+    ///   - error: the error information (optional)
+    ///   - attributes: a dictionary of attributes (optional) to add for this message. If an attribute with
+    /// the same key already exist in this logger, it will be overridden (only for this message).
+    func log(level: LogLevel, message: String, error: Error?, attributes: [String: Encodable]?)
+
+    /// General purpose logging method.
+    /// Sends a log with certain `level`, `message`, `errorKind`,  `errorMessage`,  `stackTrace` and `attributes`.
+    ///
+    /// This method is meant for non-native or cross platform frameworks (such as React Native or Flutter) to send error information
+    /// to Datadog. Although it can be used directly, it is recommended to use other methods declared on `Logger`.
+    ///
+    /// - Parameters:
+    ///   - level: the log level
+    ///   - message: the message to be logged
+    ///   - errorKind: the kind of error reported
+    ///   - errorMessage: the message attached to the error
+    ///   - stackTrace: a string representation of the error's stack trace
+    ///   - attributes: a dictionary of attributes (optional) to add for this message. If an attribute with
+    /// the same key already exist in this logger, it will be overridden (only for this message).
+    func log(
+        level: LogLevel,
+        message: String,
+        errorKind: String?,
+        errorMessage: String?,
+        stackTrace: String?,
+        attributes: [String: Encodable]?
+    )
```

Comment:
blocker/ Like the comment says: it is **only for cross-platform SDKs** in which case we should provide this method through separate `_internal` interface:
```swift
let logger: LoggerProtocol
logger._internal.log(...)
```
Part of V2, we want to strictly segregate _internal-public_ interface from _user-facing_ interface to not pollute and having to maintain backward-compatibility in the latter, endlessly.

For consistency, let's have `LoggerProtocol+Internal` like we did with [`RUMProtocol+Internal.swift`](https://github.com/DataDog/dd-sdk-ios/blob/feature/v2/DatadogRUM/Sources/RUMMonitorProtocol%2BInternal.swift).

Reply:
I'm adopting a slightly different approach because I have multiple instances that need to comply, `ConsoleLogger`, `RemoteLogger`, and `CombinedLogger`:
```swift
/// Internal Logger for Cross-Platform access.
public protocol InternalLoggerProtocol {
    /// General purpose logging method.
    /// Sends a log with certain `level`, `message`, `errorKind`,  `errorMessage`,  `stackTrace` and `attributes`.
    ///
    /// This method is meant for non-native or cross platform frameworks (such as React Native or Flutter) to send error information
    /// to Datadog. Although it can be used directly, it is recommended to use other methods declared on `Logger`.
    ///
    /// - Parameters:
    ///   - level: the log level
    ///   - message: the message to be logged
    ///   - errorKind: the kind of error reported
    ///   - errorMessage: the message attached to the error
    ///   - stackTrace: a string representation of the error's stack trace
    ///   - attributes: a dictionary of attributes (optional) to add for this message. If an attribute with
    /// the same key already exist in this logger, it will be overridden (only for this message).
    func log(
        level: LogLevel,
        message: String,
        errorKind: String?,
        errorMessage: String?,
        stackTrace: String?,
        attributes: [String: Encodable]?
    )
}

private struct NOPInternalLogger: InternalLoggerProtocol {
    func log(
        level: LogLevel,
        message: String,
        errorKind: String?,
        errorMessage: String?,
        stackTrace: String?,
        attributes: [String : Encodable]?
    ) { }
}

/// Extends `LoggerProtocol` with additional methods designed for Datadog cross-platform SDKs.
extension LoggerProtocol {
    /// Grants access to an internal interface utilized only by Datadog cross-platform SDKs.
    /// **It is not meant for public use** and it might change without prior notice.
    public var _internal: InternalLoggerProtocol {
        self as? InternalLoggerProtocol ?? NOPInternalLogger()
    }
}
```

The idea is the same but more generic 👍

Reply:
👌 Works for me as long as it's separate and non user-facing interface 💪